splash/browser_tab.py

```diff
@@ -1116,6 +1198,176 @@ def log(self, message, min_level=None):
+
+    def remove(self):
+        return self.storage.remove_event(self.id)
+
+
+class EventHandlersStorage(QObject):
+    """
+    Object that allows to store JavaScript event listeners.
+
+    This creates a JavaScript-compatible object (can be added to `window`)
+    that has `run_function()` function which is called from JS when the event
+    is triggered and the event listener is called.
+    """
+    def __init__(self, parent, events_storage):
+        self.name = get_id()
+        self.events_storage = events_storage
+        self.storage = {}
+        super(EventHandlersStorage, self).__init__(parent)
+
+    def add(self, func):
+        func_id = get_id()
+
+        event_wrapper = u"window[{storage_name}].add(event)".format(
+            storage_name=escape_js(self.events_storage.name),
+        )
+        js_func = u"window[{storage_name}][{func_id}] = " \
+                  u"function(event) {{ window[{storage_name}].run({func_id}, {event}, event) }}"\
+            .format(
+                storage_name=escape_js(self.name),
+                func_id=escape_js(func_id),
+                event=event_wrapper
+            )
+
+        escape_and_evaljs(self.parent().web_page.mainFrame(), js_func)
+
+        self.storage[func_id] = func
+        return func_id
+
+    def remove(self, func_id):
+        if self.storage.get(func_id, None) is not None:
+            del self.storage[func_id]
+
+    @pyqtSlot(str, str, 'QVariantMap', name="run")
+    def run_function(self, func_id, event_id, event):
+        wrapped_event = Event(self.events_storage, event_id, event)
+        self.storage[func_id].on_call_after.append(wrapped_event.remove)
+        self.storage[func_id](wrapped_event)
+
+
+class EventsStorage(QObject):
+    """
+    Object that allows to store JavaScript Event objects and access them.
+
+    This creates a JavaScript-compatible object (can be added to `window`)
+    that has `get_id()` function which can be called from JavaScript for
+    retrieving a unique id for each event object.
+
+    After adding to the JS window object the `init_storage(self)` method
+    should be called to initialize the storage. During the initialization
+    the storage object is connected to the QT signals which allows to call
+    appropriate methods of the specified event.
+    """
+    preventDefault = pyqtSignal(str)
+    stopImmediatePropagation = pyqtSignal(str)
+    stopPropagation = pyqtSignal(str)
+
+    def __init__(self, parent):
+        self.name = get_id()
+        super(EventsStorage, self).__init__(parent)
+
+    def init_storage(self):
+        frame = self.parent().web_page.mainFrame()
+        eval_expr = u"eval({})".format(escape_js("""
+        (function() {{
+            var storage = window[{storage_name}];
+
+            storage.events = {{}};
+
+            storage.callMethod = function(methodName) {{
+                return function(eventId) {{
+                    var storage = window[{storage_name}].events;
```

Comment:
a nitpick: it may be better to use another name, which doesn't shadow a name from the outer scope